### PR TITLE
Introduce Kernel::Weighted_point_3 together with functors

### PR DESCRIPTION
**Pull request by afabri**
See also this [small feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Weighted_point_3)
> Pre-approved by Monique 14:36, 9 December 2016 (CET)

### CODE REVIEW

**Comment by lrineau** [comment]
Please push to the pull-request, so that I can easily that the progress.

**Comment by lrineau** [comment]
Tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.9-Ic-26.shtml

I have also tested that examples from `CGAL-4.7/examples/Triangulation_3/` (and those from `Triangulation_2`) still compile with this PR.

**Comment by lrineau** [comment]
The branch has conflicts with `cgal/master`. You should consider refreshing it again.

**Comment by afabri** [comment]
refreshed.

File: Triangulation_3/include/CGAL/Regular_triangulation_3.h [hunk 1]
```diff
@@ -166,12 +174,16 @@ namespace CGAL {
...
     using Tr_Base::is_valid;
 
-    Regular_triangulation_3(const Gt & gt = Gt(), Lock_data_structure *lock_ds = NULL)
-      : Tr_Base(Tr_Base_Gt(gt), lock_ds), hidden_point_visitor(this)
+    Regular_triangulation_3()
+      : Tr_Base(Tr_Base_Gt(), NULL), hidden_point_visitor(this)
+    {}
+
+    Regular_triangulation_3(const Gt &, Lock_data_structure *lock_ds = NULL)
+      : Tr_Base(Tr_Base_Gt(), lock_ds), hidden_point_visitor(this)
```
Comment (lrineau) [inline Triangulation_3/include/CGAL/Regular_triangulation_3.h]: I do not understand that change. It the user has a custom `Gt` that carries state, you cannot ignore its value!

The left side of the patch was right: the `gt` object must be passed to the constructor of `Tr_Base_Gt`, and the constructor of `Tr_Base_Gt` must then pass `gt` to the constructor of its base class!

Comment (janetournois) [inline Triangulation_3/include/CGAL/Regular_triangulation_3.h]: We then need to add two constructors to `Regular_triangulation_euclidean_traits_3`. I fixed that in commit 2a4a41a612f8da5f0130ff043f6d1298d4512504

**Comment by lrineau** [comment]
This PR have thousands of commits, and a lot of conflicts with `master`. I would like to make it tested in `integration`, but I think the branch should be rebased, or squashed, first.

it seems the real numbers are 107 commits, with 139 modified/removed/added files. That is a lot, but a lot less than what Github current displays.

**Comment by janetournois** [comment]
I'm (trying to) rebasing it

**Comment by janetournois** [comment]
rebase done.
- `Triangulation_3` tests and examples compile (locally)
- `Mesh_3` tests do not compile, I'm working on it

**Comment by janetournois** [comment]
I compiled locally and successfully 
- tests and examples of `Triangulation_3`
- tests and examples of `Mesh_3`
- demo `Polyhedron`

I would say this PR is ready for testing!

**Comment by lrineau** [comment]
The modifications required in `Mesh_3` and `Polyhedron/demo` show that the PR breaks the compatibility. It is expected? I think it is not. `CGAL::Point_3<K>` is implicitly constructible  from `CGAL::Weighted_point_3<K>` and the reverse is also true. I am pretty sure those implicit conversion, even the surprising one that silently drops the weight of weighted point, were introduced for compatibility. So why is it not sufficient?

**Comment by janetournois** [comment]
- About `Polyhedron/demo` : the fix does not concern the weighted/bare points, but the 2nd template parameter of `CGAL::polylines_to_protect` (word type). I don't think it is directly related to this PR. The error was probably introduced during the rebase.

- About `Mesh_3`, I agree that this is a bit disturbing. I would like to discuss the technical solution with you and @afabri . It concerns commits  	137fe12 and  	19dadd6

**Comment by janetournois** [comment]
we're now using Kernel functors. There is actually no need to explicitly convert from weighted to bare point.

File: Installation/changes.html [hunk 2]
```diff
@@ -164,6 +169,25 @@ <h2 id="release4.10">Release 4.10 </h2>
...
+      Users who need this feature must use a version prior to 4.9</li>
+  </ul>
+
+ <h3>3D Triangulations</h3>
+  <ul>
+    <li><b>Breaking change</b>: Removed the class <code>Regular_triangulation_filtered_traits_3</code>.
+      Users must replace it by the filtered kernel that they passed as template argument.</li>
+    <li><b>Breaking change</b>: The number type of weighted points in regular triangulations is no longer
+        a template parameter but the field type of the geometric traits class.
+        Users who need this feature must use a version prior to 4.9.</li>
```
Comment (lrineau) [inline Installation/changes.html]: - I think you should also emphasis that `Tr::Geom_traits::Point` no longer work, and should be replaced by either `Tr::Geom_traits::Point_3` or `Tr::Geom_traits::Weighted_point_3`.

- The introduction of the `Regular_triangulation_vertex_base_3` (and `_2`), and why they should be used, should also be documented as a breaking change.

**Comment by lrineau** [comment]
@janetournois:

KDS in last tested seems [fully red].

[fully red]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-52.shtml#Kinetic_data_structures

That seems to be because of an error in this PR, related to the constructors of `Regular_triangulation_euclidean_traits_3`:
```
/usr/local/bin/clang++   -DCGAL_TEST_SUITE -DCGAL_USE_GMP -DCGAL_USE_MPFR -I/home/cgal_tester/build/src/cmake/platforms/Ubuntu-clang-latest/test/Kinetic_data_structures/include -I/home/cgal_tester/build/src/cmake/platforms/Ubuntu-clang-latest/test/Kinetic_data_structures/../../include -isystem /usr/include/x86_64-linux-gnu -I/home/cgal_tester/build/src/cmake/platforms/Ubuntu-clang-latest/test/Kinetic_data_structures -I/home/cgal_tester/build/src/cmake/platforms/Ubuntu-clang-latest/include -I/mnt/testsuite/include  -DDONT_USE_BOOST_PROGRAM_OPTIONS -Wthread-safety -g   -o CMakeFiles/random_kds.dir/random_kds.cpp.o -c /home/cgal_tester/build/src/cmake/platforms/Ubuntu-clang-latest/test/Kinetic_data_structures/random_kds.cpp
In file included from /home/cgal_tester/build/src/cmake/platforms/Ubuntu-clang-latest/test/Kinetic_data_structures/random_kds.cpp:6:
In file included from /mnt/testsuite/include/CGAL/Kinetic/Regular_triangulation_exact_simulation_traits.h:33:
/mnt/testsuite/include/CGAL/Regular_triangulation_euclidean_traits_3.h:37:3: error: constructor for 'CGAL::Regular_triangulation_euclidean_traits_3<CGAL::Kinetic::Regular_triangulation_instantaneous_kernel<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits>, CGAL::Gmpq>' must explicitly initialize the base class 'CGAL::Kinetic::Regular_triangulation_instantaneous_kernel<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits>' which does not have a default constructor
  Regular_triangulation_euclidean_traits_3(const K&) {}
  ^
/mnt/testsuite/include/CGAL/Regular_triangulation_3.h:174:17: note: in instantiation of member function 'CGAL::Regular_triangulation_euclidean_traits_3<CGAL::Kinetic::Regular_triangulation_instantaneous_kernel<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits>, CGAL::Gmpq>::Regular_triangulation_euclidean_traits_3' requested here
      : Tr_Base(Tr_Base_Gt(gt), lock_ds), hidden_point_visitor(this)
                ^
/mnt/testsuite/include/CGAL/Kinetic/internal/Triangulation_helper_3.h:42:77: note: in instantiation of member function 'CGAL::Regular_triangulation_3<CGAL::Kinetic::Regular_triangulation_instantaneous_kernel<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits>, CGAL::Triangulation_data_structure_3<CGAL::Kinetic::Regular_triangulation_vertex_base_3<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits, CGAL::Triangulation_vertex_base_3<CGAL::Kinetic::Regular_triangulation_instantaneous_kernel<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits>, CGAL::Triangulation_ds_vertex_base_3<void> > >, CGAL::Kinetic::Regular_triangulation_cell_base_3<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits, CGAL::Triangulation_cell_base_3<CGAL::Kinetic::Regular_triangulation_instantaneous_kernel<CGAL::Kinetic::Regular_triangulation_exact_simulation_traits>, CGAL::Triangulation_ds_cell_base_3<void> > >, CGAL::Sequential_tag>, CGAL::Default>::Regular_triangulation_3' requested here
  Triangulation_helper_3(typename Tri::Geom_traits gt= Tri::Geom_traits()): Tri(gt){}
                                                                            ^
```

See https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-52/Kinetic_data_structures/TestReport_gimeno_Ubuntu-clang-latest.gz for the full compilation log.

**Comment by lrineau** [comment]
@afabri:

Another compilation error, in Alpha_shape_3:
```
[ 50%] Building CXX object CMakeFiles/test_fixed_alpha_shape_3.dir/test_fixed_alpha_shape_3.cpp.o
/bin/clang++   -DCGAL_TEST_SUITE -DCGAL_USE_GMP -DCGAL_USE_MPFR -I/home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/test/Alpha_shapes_3/include -I/home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/test/Alpha_shapes_3/../../include -I/home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/test/Alpha_shapes_3 -I/home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/include -I/mnt/testsuite/include  -Wall -O3 -DCGAL_NDEBUG   -o CMakeFiles/test_fixed_alpha_shape_3.dir/test_fixed_alpha_shape_3.cpp.o -c /home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/test/Alpha_shapes_3/test_fixed_alpha_shape_3.cpp
In file included from /home/cgal_tester/build/src/cmake/platforms/ArchLinux-clang-Release/test/Alpha_shapes_3/test_fixed_alpha_shape_3.cpp:1:
/mnt/testsuite/include/CGAL/Fixed_alpha_shape_3.h:411:14: error: call to non-static member function without an object argument
    Tr_Base::remove(vertex_to_remove,remover);
    ~~~~~~~~~^~~~~~
```
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-52/Alpha_shapes_3/TestReport_lrineau_ArchLinux-clang-Release.gz

Probably the typedef `Tr_Base` is wrong, and does not correspond to the real base class.

**Comment by lrineau** [comment]
The interpolation lines are fully red too, with runtime errors:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-52.shtml#Interpolation

**Comment by lrineau** [comment]
There also a compilation error with `-DCGAL_NO_DEPRECATED_CODE`. It might be because of this PR:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-52/Alpha_shapes_2_Examples/TestReport_lrineau_Ubuntu-Latest-NO_DEPRECATED_CODE.gz
```
In file included from /home/cgal_tester/build/src/cmake/platforms/Ubuntu-Latest-NO_DEPRECATED_CODE/test/Alpha_shapes_2_Examples/ex_weighted_alpha_shapes_2.cpp:2:0:
/mnt/testsuite/include/CGAL/Weighted_alpha_shape_euclidean_traits_2.h:32:2: error: #error The class Weighted_alpha_shape_euclidean_traits_2<K> is deprecated in favor of Regular_triangulation_euclidean_traits_2<K>.
 #error The class Weighted_alpha_shape_euclidean_traits_2<K> is deprecated in favor of Regular_triangulation_euclidean_traits_2<K>.
  ^
```

**Comment by lrineau** [comment]
And same in ipelets: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.10-Ic-52/CGAL_ipelets_Demo/TestReport_lrineau_Ubuntu-Latest-NO_DEPRECATED_CODE.gz

**Comment by lrineau** [comment]
The last commit https://github.com/CGAL/cgal/pull/585/commits/2835e361c851fa9bccc3ff47e981e8ee970f0d58 should fix the runtime errors in Interpolation, and compilation errors in KDS. They both were due to the same issue: the base class of `Regular_triangulation_euclidean_traits_2` (or `_3`) can have only a non-default constructor (for example because the class carries data).

**Comment by lrineau** [comment]
I think I have fixed most of the aforementioned issues. The remaining known issue is the handling of `-DCGAL_NO_DEPRECATED_CODE` in `Alpha_shape_2`, and `CGAL_ipelets`.

**Comment by lrineau** [comment]
Those files are executable by error:
```
Triangulation_3/doc/Triangulation_3/fig/ortho.svg
Triangulation_3/doc/Triangulation_3/fig/sidedim2.svg
```

**Comment by lrineau** [comment]
This PR has not been tested since commit https://github.com/CGAL/cgal/pull/585/commits/6e098178b930c544ff1b31b7174d7f3eb877f5d7. I will re-test it tonight alone, hopefully for the last time.

**Pull request merged**